app/components/file-viewer/canvas-viewer.jsx

```diff
@@ -0,0 +1,111 @@
+import React from 'react';
+import Markdown from 'markdownz';
+import LoadingIndicator from '../loading-indicator';
+import modelSelector from '../modelling';
+import alert from '../../lib/alert';
+
+class CanvasViewer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.onLoad = this.onLoad.bind(this);
+    this.state = {
+      loading: true
+    };
+  }
+  componentDidMount() {
+    // add the canvas and prep for rendering
+    this.createNewModel(this.props);
+  }
+  shouldComponentUpdate(nextProps, nextState) {
+    // check if a new subject has been provided
+    if (this.state.loading !== nextState.loading || this.props.src !== nextProps.src) {
+      return true;
+    }
+    // Only re-render when annotation has changed, or we have zoomed/panned
+    // JSON is expensive, so reduce the test as much as possible
+    if (JSON.stringify(nextProps.annotations) !== JSON.stringify(this.props.annotations)) {
```

Comment:
Easiest way to do a deep comparison between objects

Reply:
Noting that `annotations` is an array, not an object, but I'm not sure if that makes a difference here.

Reply:
is this a preferable way of doing a deep equality check?
`const deepEquals = (a, b) => (Object.keys(a).length ? Object.keys(a).every(k => deepEquals(a[k], b[k])) : a === b);`
doesn't catch circular references but you could wrap it
```javascript
const deepEquals = (a, b) => {
  const foo = (a, b) => (Object.keys(a).length ? Object.keys(a).every(k => foo(a[k], b[k])) : a === b);
  try {
    return foo(a, b);
  } catch (RangeError) {
    throw('TypeError: Cannot check deep equality of circular structure');
    return false;
  }
}
function Foo() {
   this.circular = this;
   this.abc = "Hello";
}
const foo = new Foo();
deepEquals(foo, foo);
```

Reply:
Are you watching to see if a particular key in the current annotation changes? If so, why not just compare that key on old and new props in `componentWillReceiveProps` or `componentWillUpdate`?

Reply:
lodash's `_.isMatch()` might be what you're looking for and we already have lodash installed as a dependency: https://lodash.com/docs/4.17.4#isMatch

Reply:
I'd just compare the old and new annotation, to be honest: https://github.com/zooniverse/Panoptes-Front-End/blob/60a5ff4db7e344b0365da2d288b432b97b44ead4/app/classifier/frame-annotator.jsx#L23

Reply:
Comparing the entire annotations array seems wasteful when the only item that could have changed is the current annotation.

Reply:
I'm still getting the problem with object references not changing so would need to do a deep copy somewhere.
The frame annotator check doesn't fire either when a drawing tool is manipulated (`{value: [{...}]}`), only for tasks with basic values (`{ value: 0 }`).

Reply:
Looking more closely, I don't think that line in the frame annotator is supposed to fire when the annotation changes. I think it's there to detect when the task changes. That's confusing.

Reply:
I'm not sure if this is relevant here but #4272 fixes some bugs where annotation changes weren't correctly updating classifier state.